pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+        multiplicatively removes them from the current image.  If available, the
+        calculation is propagated to the variance image; however, no uncertainty
+        in the flat-field corrections are included.
+
+        .. warning::
+
+            If you want the spatial flexure to be accounted for, you must first
+            calculate the shift using
+            :func:`~pypeit.images.rawimage.RawImage.spatial_flexure_shift`.
 
         Args:
-            force (bool, optional):
-                Force the processing even if the image was already processed
+            flatimages (:class:`~pypeit.flatfield.FlatImages`):
+                Flat-field images used to apply flat-field corrections.
+            slits (:class:`~pypeit.slittrace.SlitTraceSet`, optional):
+                Used to construct the slit illumination profile, and only 
+                required if this is to be calculated and normalized out.  See
+                :func:`~pypeit.flatfield.FlatImages.fit2illumflat`.
+            force (:obj:`bool`, optional):
+                Force the image to be field flattened, even if the step log
+                (:attr:`steps`) indicates that it already has been.
+            debug (:obj:`bool`, optional):
+                Run in debug mode.
+        """
+        step = inspect.stack()[0][3]
+        if self.steps[step] and not force:
+            # Already field flattened
+            msgs.warn('Image was already flat fielded.')
+            return
+
+        # Check input
+        if flatimages.pixelflat_norm is None:
+            # We cannot do any flat-field correction without a pixel flat (yet)
+            msgs.error("Flat fielding desired but not generated/provided.")
+        if self.par['use_illumflat'] and slits is None:
+            msgs.error('Need to provide slits to create illumination flat.')
+        if self.par['use_specillum'] and flatimages.pixelflat_spec_illum is None:
+            msgs.error("Spectral illumination correction desired but not generated/provided.")
+
+        # Generate the illumination flat, as needed
+        illum_flat = 1.0
+        if self.par['use_illumflat']:
+            illum_flat = flatimages.fit2illumflat(slits, flexure_shift=self.spat_flexure_shift)
+            if debug:
+                left, right = slits.select_edges(flexure=self.spat_flexure_shift)
+                viewer, ch = display.show_image(illum_flat, chname='illum_flat')
+                display.show_slits(viewer, ch, left, right)  # , slits.id)
+                #
+                orig_image = self.image.copy()
+                viewer, ch = display.show_image(orig_image, chname='orig_image')
+                display.show_slits(viewer, ch, left, right)  # , slits.id)
+
+        # Apply the relative spectral illumination
+        spec_illum = 1.
+        if self.par['use_specillum']:
+            # TODO: Why is the convention for spec_illum different from the
```

Comment:
I think you are asking why is there the inverse here, i.e. why not just store the inverse instead of having to compute the inverse? I would agree with that.

Reply:
As per our zoom conversation, the convention is changed.